go/pkg/casng/batching.go

```diff
@@ -329,5 +434,216 @@ func (u *BatchingUploader) DigestTree(ctx context.Context, root impath.Absolute,
+			return
+		}
+
+		// Clear the digest to ensure proper hierarchy caching in the digester.
+		r.Digest = digest.Digest{}
+
+		if pathSeen[r.Path] {
+			continue
+		}
+		pathSeen[r.Path] = true
+		parent := r.Path.Dir()
+		skip := false
+		for ; parent.String() != execRoot.String() && parent.String() != localPrefix.String(); parent = parent.Dir() {
+			if pathSeen[parent] {
+				// First iteration, parent == r.Path, and this means a duplicate request because the list is sorted (If not sorted, it may also mean a redundant ancestor).
+				// Other iterations, it means this request is redundant because an ancestor request already exist.
+				skip = true
+				break
+			}
+		}
+		if skip {
+			continue
+		}
+		paths = append(paths, r.Path.String())
+
+		// Shift left to accumulate included requests at the beginning of the array to avoid allocating another one.
+		reqs[i] = r
+		i++
+	}
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload.tree; got=%d, uploading=%d", len(reqs), i)
+
+	// Reslice to take included (shifted) requests only.
+	reqs = reqs[:i]
+	// paths is already sorted because reqs was sorted.
+	filterID = digest.NewFromBlob([]byte(strings.Join(paths, ""))).String()
+	filterIDFunc := func() string { return filterID }
+	for _, r := range reqs {
+		r.Exclude.ID = filterIDFunc // r is a copy, but r.Exclude is a reference.
+	}
+
+	// 2nd, Upload the requests first to digest the files and cache the nodes.
+	uploaded, stats, err = u.Upload(ctx, reqs...)
+	if err != nil {
+		return
+	}
+
+	// 3rd, Compute the shared ancestor nodes and upload them.
+
+	// This block creates a flattened tree of the paths in reqs rooted at execRoot.
+	// Each key is an absolute path to a node in the tree and its value is a list of absolute paths that any of them can be a key as well.
+	// Example: /a: [/a/b /a/c], /a/b: [/a/b/foo.go], /a/c: [/a/c/bar.go]
+	dirChildren := make(map[impath.Absolute]map[impath.Absolute]proto.Message, len(pathSeen))
+	for _, r := range reqs {
+		// Each request in reqs must correspond to a cached node.
+		node := u.Node(r)
+		if node == nil {
+			err = fmt.Errorf("cannot construct the merkle tree with a missing node for path %q", r.Path)
+			return
+		}
+
+		// Add this leaf node to its ancestors.
+		rpStr := strings.TrimPrefix(r.Path.String(), execRoot.String()+string(filepath.Separator))
+		if strings.HasPrefix(rpStr, workingDir.String()) {
+			rpStr = strings.Replace(rpStr, workingDir.String(), remoteWorkingDir.String(), 1)
+		}
+		rp := execRoot.Append(impath.MustRel(rpStr))
+		// rp := remotePath[r.Path]
+		parent := rp
+		for {
+			rp = parent
+			parent = parent.Dir()
+			children := dirChildren[parent]
+			if children == nil {
+				children = make(map[impath.Absolute]proto.Message, len(reqs)) // over-allocating to avoid copying on growth.
+				dirChildren[parent] = children
+			}
+			// If the parent already has this child, then no need to traverse up the tree again.
+			_, ancestrySeen := children[rp]
+			children[rp] = node
```

Comment:
Is it intended that this is nil for every loop after the first one?

Reply:
Yes. This looks more concise than using a boolean.